ledger/src/bank_forks.rs

```diff
@@ -177,35 +177,50 @@ impl BankForks {
-
-        // Generate a snapshot if snapshots are configured and it's been an appropriate number
-        // of banks since the last snapshot
+        // Generate each snapshot at a fixed interval
         if self.snapshot_config.is_some() && snapshot_package_sender.is_some() {
             let config = self.snapshot_config.as_ref().unwrap();
-            info!("setting snapshot root: {}", root);
-            if root - self.last_snapshot_slot >= config.snapshot_interval_slots as Slot {
-                let mut snapshot_time = Measure::start("total-snapshot-ms");
-                let r = self.generate_snapshot(
-                    root,
-                    &root_bank.src.roots(),
-                    snapshot_package_sender.as_ref().unwrap(),
-                    snapshot_utils::get_snapshot_archive_path(&config.snapshot_package_output_path),
-                );
-                if r.is_err() {
-                    warn!("Error generating snapshot for bank: {}, err: {:?}", root, r);
-                } else {
-                    self.last_snapshot_slot = root;
+            let mut banks = vec![root_bank];
+            let parents = root_bank.parents();
+            banks.extend(parents.iter());
+            for bank in banks {
+                let parent_slot = bank.parent_slot();
+                let root = bank.slot();
+                if Self::check_interval(root, parent_slot, config.snapshot_interval_slots) {
+                    // Generate a snapshot if snapshots are configured and it's been an appropriate number
+                    // of banks since the last snapshot
+                    info!(
+                        "setting snapshot root: {} interval: {}",
+                        root, config.snapshot_interval_slots
+                    );
+                    if root - self.last_snapshot_slot >= config.snapshot_interval_slots as Slot {
+                        bank.squash();
+                        let mut snapshot_time = Measure::start("total-snapshot-ms");
+                        let r = self.generate_snapshot(
+                            root,
+                            &bank.src.roots(),
+                            snapshot_package_sender.as_ref().unwrap(),
+                            snapshot_utils::get_snapshot_archive_path(
+                                &config.snapshot_package_output_path,
+                            ),
+                        );
+                        if r.is_err() {
+                            warn!("Error generating snapshot for bank: {}, err: {:?}", root, r);
+                        } else {
+                            self.last_snapshot_slot = root;
+                        }
+
+                        // Cleanup outdated snapshots
+                        self.purge_old_snapshots();
+                        snapshot_time.stop();
+                        inc_new_counter_info!("total-snapshot-ms", snapshot_time.as_ms() as usize);
+                    }
+                    break;
```

Comment:
I think another loop to find the most latest slot is preferable rather than the first and oldest un-`squashed` slot, considering this corner case: https://github.com/solana-labs/solana/issues/8156

Reply:
parents are in newest first.  so we should find the latest slot.

Reply:
I see!